packages/next/src/shared/lib/router/router.ts

```diff
@@ -1096,7 +1096,9 @@ export default class Router implements BaseRouter {
             // a hard navigation
             if (matchesBflStatic || matchesBflDynamic) {
               handleHardNavigation({
-                url: addBasePath(addLocale(as, locale || this.locale)),
+                url: addBasePath(
+                  addLocale(as, locale || this.locale, this.defaultLocale)
```

Comment:
`addLocale` skips adding the locale if `locale` is equal to `defaultLocale`:

https://github.com/vercel/next.js/blob/02125cf3b1dfba52b240fd6e0f959d896e4b6195/packages/next/src/shared/lib/router/utils/add-locale.ts#L15-L23